File: code-recipes/java/src/main/java/easyship/CalculateOrderDimensionsRecipe.java
package easyship;

import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.catalogitems.v2022_04_01.CatalogApi;
import software.amazon.spapi.api.listings.items.v2021_08_01.ListingsApi;
import software.amazon.spapi.api.orders.v0.OrdersV0Api;
import software.amazon.spapi.models.orders.v0.GetOrderItemsResponse;
import software.amazon.spapi.models.orders.v0.OrderItem;
import util.Constants;
import util.Recipe;

import com.amazon.SellingPartnerAPIAA.LWAException;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

/**
 * Code Recipe to calculate order weight and dimensions for EasyShip
 * Steps:
 * 1. Get order items to retrieve all SKUs
 * 2. For each SKU, search Catalog Items API for ASIN dimensions
 * 3. Fallback to Listings Items API if dimensions not available
 * 4. Sum all weights and dimensions
 * 
 * IMPORTANT: Neither APIs guarantees 100% of data availability.
 */
public class CalculateOrderDimensionsRecipe extends Recipe {

    private OrdersV0Api ordersApi;
    private CatalogApi catalogApi;
    private ListingsApi listingsApi;
    private String amazonOrderId;
    private String marketplaceId;
    private String sellerId;

    @Override
    protected void start() {
        setupOrderDetails();
        initializeApis();
        GetOrderItemsResponse orderItemsResponse = getOrderItems();
        calculateTotalDimensions(orderItemsResponse);
        System.out.println("✅ Successfully calculated order dimensions");
    }

    private void setupOrderDetails() {
        // IMPORTANT: Replace these sample values with actual values from your environment
        // These IDs are for demonstration purposes only
        amazonOrderId = "702-3035602-4225066";
        marketplaceId = "A1AM78C64UM0Y8";
        sellerId = "A2ZPJ4TLUOSWY8";
        System.out.println("Order details configured: " + amazonOrderId);
    }

    private void initializeApis() {
        ordersApi = new OrdersV0Api.Builder()
                .lwaAuthorizationCredentials(lwaCredentials)
                .endpoint(Constants.BACKEND_URL)
                .disableAccessTokenCache()
                .build();
        
        catalogApi = new CatalogApi.Builder()
                .lwaAuthorizationCredentials(lwaCredentials)
                .endpoint(Constants.BACKEND_URL)
                .disableAccessTokenCache()
                .build();
        
        listingsApi = new ListingsApi.Builder()
                .lwaAuthorizationCredentials(lwaCredentials)
                .endpoint(Constants.BACKEND_URL)
                .disableAccessTokenCache()
                .build();
        
        System.out.println("APIs initialized");
    }

    private GetOrderItemsResponse getOrderItems() {
        try {
            GetOrderItemsResponse response = ordersApi.getOrderItems(amazonOrderId, null);
            System.out.println("Order items retrieved: " + response.getPayload().getOrderItems().size() + " items");
            return response;
        } catch (ApiException | LWAException e) {
            throw new RuntimeException("Failed to get order items", e);
        }
    }

    private void calculateTotalDimensions(GetOrderItemsResponse orderItemsResponse) {
        BigDecimal totalWeight = BigDecimal.ZERO;
        BigDecimal totalLength = BigDecimal.ZERO;
        BigDecimal totalWidth = BigDecimal.ZERO;
        BigDecimal totalHeight = BigDecimal.ZERO;
        
        for (OrderItem orderItem : orderItemsResponse.getPayload().getOrderItems()) {
            String sku = orderItem.getSellerSKU();
            String asin = orderItem.getASIN();
            int quantity = orderItem.getQuantityOrdered();
            
            System.out.println("Processing SKU: " + sku + ", ASIN: " + asin + ", Quantity: " + quantity);
            
            Map<String, Object> attributes = getDimensionsFromCatalog(asin);
            if (attributes == null) {
                attributes = getDimensionsFromListings(sku);
            }
            
            if (attributes != null) {
                Map<String, BigDecimal> dimensions = extractPackageDimensions(attributes);
                BigDecimal weight = extractPackageWeight(attributes);
                
                totalWeight = totalWeight.add(weight.multiply(BigDecimal.valueOf(quantity)));
                totalLength = totalLength.add(dimensions.get("length").multiply(BigDecimal.valueOf(quantity)));
                totalWidth = totalWidth.add(dimensions.get("width").multiply(BigDecimal.valueOf(quantity)));
                totalHeight = totalHeight.add(dimensions.get("height").multiply(BigDecimal.valueOf(quantity)));
            }
        }
        
        System.out.println("\nTotal Order Dimensions:");
        System.out.println("  Total Length: " + totalLength + " cm");
        System.out.println("  Total Width: " + totalWidth + " cm");
        System.out.println("  Total Height: " + totalHeight + " cm");
        System.out.println("  Total Weight: " + totalWeight + " g");
    }
    
    private Map<String, Object> getDimensionsFromCatalog(String asin) {
        try {
            software.amazon.spapi.models.catalogitems.v2022_04_01.Item catalogItem = 
                catalogApi.getCatalogItem(asin, List.of(marketplaceId), List.of("attributes"), null);
            System.out.println("  Retrieved dimensions from Catalog API");
            if (catalogItem.getAttributes() != null) {
                return (Map<String, Object>) catalogItem.getAttributes();
            }
            return null;
        } catch (ApiException | LWAException e) {
            System.out.println("  Catalog API failed, will try Listings API");
            return null;
        }
    }
    
    private Map<String, Object> getDimensionsFromListings(String sku) {
        try {
            software.amazon.spapi.models.listings.items.v2021_08_01.Item listingItem = 
                listingsApi.getListingsItem(sellerId, sku, List.of(marketplaceId), null, List.of("attributes"));
            System.out.println("  Retrieved dimensions from Listings API");
            return listingItem.getAttributes();
        } catch (ApiException | LWAException e) {
            System.out.println("  Failed to retrieve dimensions from Listings API");
            return null;
        }
    }

    private Map<String, BigDecimal> extractPackageDimensions(Map<String, Object> attributes) {
        List<?> packageDimensions = (List<?>) attributes.get("item_package_dimensions");
        if (packageDimensions != null && !packageDimensions.isEmpty()) {
            Map<?, ?> dims = (Map<?, ?>) packageDimensions.get(0);
            return Map.of(
                "length", extractValue((Map<?, ?>) dims.get("length")),
                "width", extractValue((Map<?, ?>) dims.get("width")),
                "height", extractValue((Map<?, ?>) dims.get("height"))
            );
        }
        return Map.of("length", BigDecimal.ZERO, "width", BigDecimal.ZERO, "height", BigDecimal.ZERO);
    }

    private BigDecimal extractPackageWeight(Map<String, Object> attributes) {
        List<?> packageWeight = (List<?>) attributes.get("item_package_weight");
        if (packageWeight != null && !packageWeight.isEmpty()) {
            Map<?, ?> weight = (Map<?, ?>) packageWeight.get(0);
            return extractValue(weight);
        }
        return BigDecimal.ZERO;
    }

    private BigDecimal extractValue(Map<?, ?> map) {
        if (map != null && map.containsKey("value")) {
            Object value = map.get("value");
            if (value instanceof Number) {
                return BigDecimal.valueOf(((Number) value).doubleValue());
            }
        }
        return BigDecimal.ZERO;
    }
}

File: code-recipes/php/src/easyship/CalculateOrderDimensionsRecipe.php
<?php

namespace Src\easyship;

use SpApi\Api\catalogItems\v2022_04_01\CatalogApi;
use SpApi\Api\listings\items\v2021_08_01\ListingsApi;
use SpApi\Api\orders\v0\OrdersV0Api;
use Src\util\Recipe;

/**
 * Code Recipe to calculate order weight and dimensions for EasyShip
 * Steps:
 * 1. Get order items to retrieve all SKUs
 * 2. For each SKU, search Catalog Items API for ASIN dimensions
 * 3. Fallback to Listings Items API if dimensions not available
 * 4. Sum all weights and dimensions
 * 
 * IMPORTANT: Neither APIs guarantees 100% of data availability.
 */
class CalculateOrderDimensionsRecipe extends Recipe
{
    private OrdersV0Api $ordersApi;
    private CatalogApi $catalogApi;
    private ListingsApi $listingsApi;
    private string $amazonOrderId;
    private string $marketplaceId;
    private string $sellerId;

    public function start(): void
    {
        $this->setupOrderDetails();
        $this->initializeApis();
        $orderItemsResponse = $this->getOrderItems();
        $this->calculateTotalDimensions($orderItemsResponse);
        echo "✅ Successfully calculated order dimensions\n";
    }

    private function setupOrderDetails(): void
    {
        // IMPORTANT: Replace these sample values with actual values from your environment
        // These IDs are for demonstration purposes only
        $this->amazonOrderId = "702-3035602-4225066";
        $this->marketplaceId = "A1AM78C64UM0Y8";
        $this->sellerId = "A2ZPJ4TLUOSWY8";
        echo "Order details configured: {$this->amazonOrderId}\n";
    }

    private function initializeApis(): void
    {
        $this->ordersApi = new OrdersV0Api($this->config);
        $this->catalogApi = new CatalogApi($this->config);
        $this->listingsApi = new ListingsApi($this->config);
        echo "APIs initialized\n";
    }

    private function getOrderItems()
    {
        $response = $this->ordersApi->getOrderItems($this->amazonOrderId);
        $itemCount = count($response->getPayload()->getOrderItems());
        echo "Order items retrieved: {$itemCount} items\n";
        return $response;
    }

    private function calculateTotalDimensions($orderItemsResponse): void
    {
        $totalWeight = 0.0;
        $totalLength = 0.0;
        $totalWidth = 0.0;
        $totalHeight = 0.0;
        
        foreach ($orderItemsResponse->getPayload()->getOrderItems() as $orderItem) {
            $sku = $orderItem->getSellerSku();
            $asin = $orderItem->getAsin();
            $quantity = $orderItem->getQuantityOrdered();
            
            echo "Processing SKU: {$sku}, ASIN: {$asin}, Quantity: {$quantity}\n";
            
            $attributes = $this->getDimensionsFromCatalog($asin);
            if ($attributes === null) {
                $attributes = $this->getDimensionsFromListings($sku);
            }
            
            if ($attributes !== null) {
                $dimensions = $this->extractPackageDimensions($attributes);
                $weight = $this->extractPackageWeight($attributes);
                
                $totalWeight += $weight * $quantity;
                $totalLength += $dimensions['length'] * $quantity;
                $totalWidth += $dimensions['width'] * $quantity;
                $totalHeight += $dimensions['height'] * $quantity;
            }
        }
        
        echo "\nTotal Order Dimensions:\n";
        echo "  Total Length: {$totalLength} cm\n";
        echo "  Total Width: {$totalWidth} cm\n";
        echo "  Total Height: {$totalHeight} cm\n";
        echo "  Total Weight: {$totalWeight} g\n";
    }
    
    private function getDimensionsFromCatalog(string $asin): ?array
    {
        try {
            $catalogItem = $this->catalogApi->getCatalogItem(
                $asin,
                [$this->marketplaceId],
                ['attributes']
            );
            echo "  Retrieved dimensions from Catalog API\n";
            return $catalogItem->getAttributes() ? (array)$catalogItem->getAttributes() : null;
        } catch (\Exception $e) {
            echo "  Catalog API failed, will try Listings API\n";
            return null;
        }
    }
    
    private function getDimensionsFromListings(string $sku): ?array
    {
        try {
            $listingItem = $this->listingsApi->getListingsItem(
                $this->sellerId,
                $sku,
                [$this->marketplaceId],
                null,
                ['attributes']
            );
            echo "  Retrieved dimensions from Listings API\n";
            return $listingItem->getAttributes();
        } catch (\Exception $e) {
            echo "  Failed to retrieve dimensions from Listings API\n";
            return null;
        }
    }

    private function extractPackageDimensions(array $attributes): array
    {
        $packageDimensions = $attributes['item_package_dimensions'] ?? null;
        if ($packageDimensions && !empty($packageDimensions)) {
            $dims = (array)$packageDimensions[0];
            return [
                'length' => $this->extractValue((array)($dims['length'] ?? [])),
                'width' => $this->extractValue((array)($dims['width'] ?? [])),
                'height' => $this->extractValue((array)($dims['height'] ?? []))
            ];
        }
        return ['length' => 0, 'width' => 0, 'height' => 0];
    }

    private function extractPackageWeight(array $attributes): float
    {
        $packageWeight = $attributes['item_package_weight'] ?? null;
        if ($packageWeight && !empty($packageWeight)) {
            return $this->extractValue((array)$packageWeight[0]);
        }
        return 0.0;
    }

    private function extractValue(array $data): float
    {
        return isset($data['value']) ? (float)$data['value'] : 0.0;
    }
}
